### PR TITLE
feat(check:policy): Verify packages have an exports field in package.json

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR adds a policy handler that checks if a package has an [exports](https://nodejs.org/api/packages.html#package-entry-pointstype) field defined. If it does, it further checks that the values in the `import` and `require` sections match the `module` and `main` values.

It also provides a resolver that adds an exports field for the package. It uses the values from the main/module/types fields to build the exports field. If a package doesn't have a module field, then only the `require` (CommonJS) section is generated.

The resolver attempts to pick correct values for CJS- or ESM-only packages, as well as those that build both

I pre-emptively adjusted the fluid-build config to exclude this handler for most of the repo. It only applies to azure and build-tools right now. As we want to include more packages, we can remove the exclusions from the config.

Note that this policy won't apply until we release and integrate build-tools with the client release group.

### CODE REVIEW

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 1]
```diff
@@ -1156,6 +1156,87 @@ export const handlers: Handler[] = [
...
+
+			if (json.types === undefined && json.typings === undefined) {
+				throw new Error(
+					"The 'types' and 'typings' field are both undefined. At least one must be defined.",
+				);
+			}
+
+			if (json.main === undefined) {
+				throw new Error("The 'main' field is undefined. It must have a value.");
+			}
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:1185]: These checks should probably be in the resolver only.

Comment (tylerbutler) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:1185]: These checks are already done in shouldCheckExportsField, so they can be removed.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 2]
```diff
@@ -1156,6 +1156,87 @@ export const handlers: Handler[] = [
...
+
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:1186]:
```suggestion
```

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 3]
```diff
@@ -1156,6 +1156,77 @@ export const handlers: Handler[] = [
...
+			// CJS exports
+			const requireField =
+				exportsRoot?.require?.default === undefined
+					? undefined
+					: normalizePathField(exportsRoot?.require?.default);
+			const mainField = normalizePathField(json.main);
+			if (requireField !== mainField) {
+				return `Incorrect 'require' extry in 'exports' field in package.json. Expected '${mainField}', got '${requireField}'`;
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:1198]:
```suggestion
				return `Incorrect 'require' entry in 'exports' field in package.json. Expected '${mainField}', got '${requireField}'`;
```

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 4]
```diff
@@ -1156,6 +1156,77 @@ export const handlers: Handler[] = [
...
+
+			// ESM exports
+			const importField =
+				exportsRoot?.import?.default === undefined
+					? undefined
+					: normalizePathField(exportsRoot?.import?.default);
+			if (json.module !== undefined) {
+				const moduleField = normalizePathField(json.module);
+				if (importField !== moduleField) {
+					return `Incorrect 'import' extry in 'exports' field in package.json. Expected '${moduleField}', got '${importField}'`;
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:1209]:
```suggestion
					return `Incorrect 'import' entry in 'exports' field in package.json. Expected '${moduleField}', got '${importField}'`;
```

**Review by alexvy86 — APPROVED** [review]
Looks good. Approving even though we might still want to discuss whether having the `"."` node under `exports` is what we want.

Also FYI in case you run into this, there's a few packages that we moved to ESM only, but I think they're packages that aren't supposed to be imported _from_ so they don't declare entrypoints or types info. `packages/test/test-end-to-end-tests` is one, and I don't recall the other(s) (@Abe27342 ?). I expect this shouldn't affect them, but thought to bring it up for reference.

**Comment by Abe27342** [comment]
> Looks good. Approving even though we might still want to discuss whether having the `"."` node under `exports` is what we want.
> 
> Also FYI in case you run into this, there's a few packages that we moved to ESM only, but I think they're packages that aren't supposed to be imported _from_ so they don't declare entrypoints or types info. `packages/test/test-end-to-end-tests` is one, and I don't recall the other(s) (@Abe27342 ?). I expect this shouldn't affect them, but thought to bring it up for reference.

@fluidframework/test-version-utils is one of them that does get imported from. The rest are all leaf e2e test/example packages I believe. The reasoning there is that we need to make sure we have the right set of compat packages installed, but mocha doesn't run top-level setup consistently before test discovery (old bug, doesn't seem like it's going to change anytime soon). So to make this available in test discovery, which enables a more convenient API, we use top-level await to guarantee that.

**Comment by tylerbutler** [comment]
> @fluidframework/test-version-utils is one of them that does get imported from. The rest are all leaf e2e test/example packages I believe. The reasoning there is that we need to make sure we have the right set of compat packages installed, but mocha doesn't run top-level setup consistently before test discovery (old bug, doesn't seem like it's going to change anytime soon). So to make this available in test discovery, which enables a more convenient API, we use top-level await to guarantee that.

@Abe27342 @alexvy86 For ESM-only packages, what should the exports entry look like? Should I just omit the "require" entry? I think I can use the `"type": "module"` field to determine if it's ESM only.

**Pull request merged**